lib/ClangImporter/ImportDecl.cpp

```diff
@@ -4535,18 +4535,45 @@ namespace {
     }
 
     template <typename T, typename U>
-    T *resolveSwiftDeclImpl(const U *decl, Identifier name, ModuleDecl *overlay) {
+    T *resolveSwiftDeclImpl(const U *decl, Identifier name, bool hasCustomName,
+                            ModuleDecl *overlay) {
       const auto &languageVersion =
           Impl.SwiftContext.LangOpts.EffectiveLanguageVersion;
 
+      // None = not enough information
+      // true/false = definitive answer
+      auto isMatch = [&](const T *singleResult) -> Optional<bool> {
+        const DeclAttributes &attrs = singleResult->getAttrs();
+
+        // Skip versioned variants.
+        if (attrs.isUnavailableInSwiftVersion(languageVersion))
+          return false;
+
+        // If Clang decl has a custom Swift name, then `name` is that name.
+        if (hasCustomName)
+          return None;
+
+        // Skip if a different name is used for Objective-C.
+        if (auto objcAttr = attrs.getAttribute<ObjCAttr>())
+          if (auto objcName = objcAttr->getName())
+            return objcName->getSimpleName() == name;
+
+        return None;
```

Comment:
…and this one? If there's no explicit ObjC attribute, can't we assume we know enough about the name to make a definitive statement?

Reply:
Instead of returning a tristate, how about passing in another argument saying whether the class's basename is already known to match? Or even just checking that up front?

Reply:
I’ve gotten rid of the tristate so the code should make a lot more sense now.

Reply:
Types need to be skipped over where the name matches but the type isn’t exposed to Objective-C. That was happening from the caller side before.